src/nft/nft.cc

```diff
@@ -365,28 +363,36 @@ State Nft::insert_word_by_parts(const State source, const std::vector<Word> &wor
     };
 
     // Add transition source --> inner_state.
-    Level inner_lvl = (num_of_levels == 1 ) ? 0 : 1;
+    Level inner_lvl = (num_of_levels == 1 ) ? 0 : (from_to_level + 1) % static_cast<Level>(num_of_levels);
     State inner_state = add_state_with_level(inner_lvl);
-    delta.add(source, get_next_symbol(0), inner_state);
+    delta.add(source, get_next_symbol(from_to_level), inner_state);
 
     // Add transition inner_state --> inner_state
     State prev_state = inner_state;
     Level prev_lvl = inner_lvl;
-    for (size_t symbol_idx{ 1 }; symbol_idx < word_total_len - 1; symbol_idx++) {
-        inner_lvl = (prev_lvl + 1) % static_cast<Level>(num_of_levels);
-        inner_state = add_state_with_level(inner_lvl);
-        delta.add(prev_state, get_next_symbol(prev_lvl), inner_state);
-        prev_state = inner_state;
-        prev_lvl = inner_lvl;
+    const size_t max_word_part_len = std::max_element(
+        word_parts_on_levels.begin(),
+        word_parts_on_levels.end(),
+        [](const Word& a, const Word& b) { return a.size() < b.size(); }
+    )->size();
+    const size_t word_total_len = num_of_levels * max_word_part_len;
+    if (word_total_len != 0) {
```

Comment:
So, what will happen when `word_total_len == 0`? Will it insert an epsilon transition from `source` to `target`?

Reply:
`word_total_len == 0` means that we insert a word which in string solving represents a dropping a regex, and if it happens that the regex is an empty word, too, we just replace epsilon with epsilon. This is an edge case that the string solving algorithms uncovered. Just below the if statement there is a `delta.add()` which will add the missing `inner_state` to `target` over epsilon (thanks to `get_next_symbol`). Since the target is some arbitrary state, we cannot just skip this epsilon transitions entirely.